// @defindex/sdk/dist/types/vault.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VaultRoles = exports.VaultSetRoleMethods = exports.VaultGetRoleMethods = exports.VaultInfoInvocationMethods = exports.VaultMethods = void 0;
/* Comprehensive Vault Methods Enum - Based on Contract Analysis */
var VaultMethods;
(function (VaultMethods) {
    // Core Vault Operations (VaultTrait)
    VaultMethods["DEPOSIT"] = "deposit";
    VaultMethods["WITHDRAW"] = "withdraw";
    VaultMethods["RESCUE"] = "rescue";
    VaultMethods["PAUSE_STRATEGY"] = "pause_strategy";
    VaultMethods["UNPAUSE_STRATEGY"] = "unpause_strategy";
    VaultMethods["GET_ASSETS"] = "get_assets";
    VaultMethods["FETCH_TOTAL_MANAGED_FUNDS"] = "fetch_total_managed_funds";
    VaultMethods["GET_ASSET_AMOUNTS_PER_SHARES"] = "get_asset_amounts_per_shares";
    VaultMethods["GET_FEES"] = "get_fees";
    VaultMethods["REPORT"] = "report";
    // Admin Interface Methods (AdminInterfaceTrait)
    VaultMethods["SET_FEE_RECEIVER"] = "set_fee_receiver";
    VaultMethods["GET_FEE_RECEIVER"] = "get_fee_receiver";
    VaultMethods["SET_MANAGER"] = "set_manager";
    VaultMethods["GET_MANAGER"] = "get_manager";
    VaultMethods["SET_EMERGENCY_MANAGER"] = "set_emergency_manager";
    VaultMethods["GET_EMERGENCY_MANAGER"] = "get_emergency_manager";
    VaultMethods["SET_REBALANCE_MANAGER"] = "set_rebalance_manager";
    VaultMethods["GET_REBALANCE_MANAGER"] = "get_rebalance_manager";
    VaultMethods["UPGRADE"] = "upgrade";
    // Vault Management Methods (VaultManagementTrait)
    VaultMethods["REBALANCE"] = "rebalance";
    VaultMethods["LOCK_FEES"] = "lock_fees";
    VaultMethods["RELEASE_FEES"] = "release_fees";
    VaultMethods["DISTRIBUTE_FEES"] = "distribute_fees";
    VaultMethods["TOTAL_SUPPLY"] = "total_supply";
    VaultMethods["BALANCE"] = "balance";
    VaultMethods["NAME"] = "name";
    VaultMethods["SYMBOL"] = "symbol";
})(VaultMethods || (exports.VaultMethods = VaultMethods = {}));
var VaultInfoInvocationMethods;
(function (VaultInfoInvocationMethods) {
    VaultInfoInvocationMethods["GET_ASSETS"] = "get_assets";
    VaultInfoInvocationMethods["FETCH_TOTAL_MANAGED_FUNDS"] = "fetch_total_managed_funds";
    VaultInfoInvocationMethods["GET_FEES"] = "get_fees";
    VaultInfoInvocationMethods["GET_MANAGER"] = "get_manager";
    VaultInfoInvocationMethods["GET_EMERGENCY_MANAGER"] = "get_emergency_manager";
    VaultInfoInvocationMethods["GET_REBALANCE_MANAGER"] = "get_rebalance_manager";
    VaultInfoInvocationMethods["GET_FEE_RECEIVER"] = "get_fee_receiver";
    VaultInfoInvocationMethods["NAME"] = "name";
    VaultInfoInvocationMethods["SYMBOL"] = "symbol";
})(VaultInfoInvocationMethods || (exports.VaultInfoInvocationMethods = VaultInfoInvocationMethods = {}));
var VaultGetRoleMethods;
(function (VaultGetRoleMethods) {
    VaultGetRoleMethods["GET_EMERGENCY_MANAGER"] = "get_emergency_manager";
    VaultGetRoleMethods["GET_REBALANCE_MANAGER"] = "get_rebalance_manager";
    VaultGetRoleMethods["GET_MANAGER"] = "get_manager";
    VaultGetRoleMethods["GET_FEE_RECEIVER"] = "get_fee_receiver";
})(VaultGetRoleMethods || (exports.VaultGetRoleMethods = VaultGetRoleMethods = {}));
var VaultSetRoleMethods;
(function (VaultSetRoleMethods) {
    VaultSetRoleMethods["SET_MANAGER"] = "set_manager";
    VaultSetRoleMethods["SET_EMERGENCY_MANAGER"] = "set_emergency_manager";
    VaultSetRoleMethods["SET_REBALANCE_MANAGER"] = "set_rebalance_manager";
    VaultSetRoleMethods["SET_FEE_RECEIVER"] = "set_fee_receiver";
})(VaultSetRoleMethods || (exports.VaultSetRoleMethods = VaultSetRoleMethods = {}));
var VaultRoles;
(function (VaultRoles) {
    VaultRoles["MANAGER"] = "manager";
    VaultRoles["EMERGENCY_MANAGER"] = "emergency-manager";
    VaultRoles["REBALANCE_MANAGER"] = "rebalance-manager";
    VaultRoles["FEE_RECEIVER"] = "fee-receiver";
})(VaultRoles || (exports.VaultRoles = VaultRoles = {}));
